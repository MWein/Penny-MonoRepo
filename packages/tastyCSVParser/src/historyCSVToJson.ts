import { verify } from './verify'

export type TastyHistory = {
  date: Date,
  underlying: string,
  side: 'Buy' | 'Sell',
  action: 'Open' | 'Close',
  quantity: number,
  expiration: string,
  strike: number,
  optionType: 'Call' | 'Put',
  price: number,
  fees: number,
  amount: number,
}


const reducer = (row): TastyHistory => {
  return {
    date: new Date(row['Date/Time']),
    underlying: row.Symbol,
    side: row['Buy/Sell'],
    action: row['Open/Close'],
    quantity: Number(row.Quantity),
    expiration: new Date(row['Expiration Date']).toISOString().split('T')[0],
    strike: Number(row.Strike),
    optionType: row['Call/Put'] === 'C' ? 'Call' : 'Put',
    price: Number(row.Price),
    fees: Number(row.Fees),
    amount: Number(row.Amount),
  }
}


export const historyCSVToJson = (csvData: string): TastyHistory[] => {
  if (!verify(csvData)) {
    return []
  }

  const csvHeader = csvData.slice(0, csvData.indexOf('\n')).split(',')
  return csvData.slice(csvData.indexOf('\n') + 1).split('\n')
    .map(row => row.split(','))
    .map(row => {
      const obj = csvHeader.reduce((object, header, index) => {
        object[header] = row[index]
        return object
      }, {})
      return obj
    })
    .filter(row => row['Transaction Code'] === 'Trade')
    .map(row => reducer(row))
}