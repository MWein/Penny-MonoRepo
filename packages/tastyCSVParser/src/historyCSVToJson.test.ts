import { historyCSVToJson } from './historyCSVToJson'
import fs from 'fs'
import path from 'path'


describe('historyCSVToJson', () => {
  it('Returns empty array if given the wrong kind of csv', async () => {
    const csvText = await fs.readFileSync(path.resolve(__dirname, '../mockCSVs/positions.csv'), 'utf-8')
    expect(historyCSVToJson(csvText)).toEqual([])
  })

  it('Returns empty array if there are no rows', async () => {
    const csvText = await fs.readFileSync(path.resolve(__dirname, '../mockCSVs/historyEmpty.csv'), 'utf-8')
    expect(historyCSVToJson(csvText)).toEqual([])
  })

  it('Parses history csv, ignores non-Trade rows', async () => {
    const csvText = await fs.readFileSync(path.resolve(__dirname, '../mockCSVs/history.csv'), 'utf-8')
    expect(historyCSVToJson(csvText)).toMatchSnapshot()
  })
})