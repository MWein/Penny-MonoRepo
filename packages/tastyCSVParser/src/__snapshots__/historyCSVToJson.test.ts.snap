// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`historyCSVToJson Parses history csv, ignores non-Trade rows 1`] = `
Array [
  Object {
    "action": "Open",
    "amount": 15,
    "date": 2022-07-13T16:19:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.15,
    "quantity": 1,
    "side": "Sell",
    "strike": 52.5,
    "underlying": "FAST",
  },
  Object {
    "action": "Close",
    "amount": -22,
    "date": 2022-07-13T15:54:00.000Z,
    "expiration": "08/19/2022",
    "fees": 0.13,
    "optionType": "Call",
    "price": 0.22,
    "quantity": 1,
    "side": "Buy",
    "strike": 30,
    "underlying": "ASAN",
  },
  Object {
    "action": "Close",
    "amount": -5,
    "date": 2022-07-13T15:50:00.000Z,
    "expiration": "08/19/2022",
    "fees": 0.13,
    "optionType": "Call",
    "price": 0.05,
    "quantity": 1,
    "side": "Buy",
    "strike": 57.5,
    "underlying": "FAST",
  },
  Object {
    "action": "Open",
    "amount": 49,
    "date": 2022-07-12T15:13:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.49,
    "quantity": 1,
    "side": "Sell",
    "strike": 45,
    "underlying": "FAST",
  },
  Object {
    "action": "Open",
    "amount": 16,
    "date": 2022-07-12T15:13:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.16,
    "quantity": 1,
    "side": "Sell",
    "strike": 57.5,
    "underlying": "FAST",
  },
  Object {
    "action": "Open",
    "amount": 23,
    "date": 2022-07-11T17:47:00.000Z,
    "expiration": "08/05/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.23,
    "quantity": 1,
    "side": "Sell",
    "strike": 55,
    "underlying": "XLE",
  },
  Object {
    "action": "Open",
    "amount": 9,
    "date": 2022-07-11T16:17:00.000Z,
    "expiration": "08/12/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.09,
    "quantity": 1,
    "side": "Sell",
    "strike": 17,
    "underlying": "NCLH",
  },
  Object {
    "action": "Open",
    "amount": 12,
    "date": 2022-07-11T16:17:00.000Z,
    "expiration": "08/12/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.12,
    "quantity": 1,
    "side": "Sell",
    "strike": 6.5,
    "underlying": "NCLH",
  },
  Object {
    "action": "Open",
    "amount": 29,
    "date": 2022-07-11T15:29:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.29,
    "quantity": 1,
    "side": "Sell",
    "strike": 38,
    "underlying": "HAL",
  },
  Object {
    "action": "Open",
    "amount": 38,
    "date": 2022-07-11T15:29:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.38,
    "quantity": 1,
    "side": "Sell",
    "strike": 22,
    "underlying": "HAL",
  },
  Object {
    "action": "Open",
    "amount": 4,
    "date": 2022-07-11T15:03:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.04,
    "quantity": 1,
    "side": "Sell",
    "strike": 20,
    "underlying": "KMI",
  },
  Object {
    "action": "Open",
    "amount": 10,
    "date": 2022-07-11T15:03:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.1,
    "quantity": 1,
    "side": "Sell",
    "strike": 13.5,
    "underlying": "KMI",
  },
  Object {
    "action": "Open",
    "amount": 83,
    "date": 2022-07-11T14:54:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.83,
    "quantity": 1,
    "side": "Sell",
    "strike": 95,
    "underlying": "PYPL",
  },
  Object {
    "action": "Open",
    "amount": 81,
    "date": 2022-07-11T14:54:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.81,
    "quantity": 1,
    "side": "Sell",
    "strike": 50,
    "underlying": "PYPL",
  },
  Object {
    "action": "Open",
    "amount": 28,
    "date": 2022-07-11T14:54:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.28,
    "quantity": 1,
    "side": "Sell",
    "strike": 24,
    "underlying": "SLB",
  },
  Object {
    "action": "Open",
    "amount": 26,
    "date": 2022-07-11T14:54:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.26,
    "quantity": 1,
    "side": "Sell",
    "strike": 44,
    "underlying": "SLB",
  },
  Object {
    "action": "Open",
    "amount": 8,
    "date": 2022-07-11T14:45:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.08,
    "quantity": 1,
    "side": "Sell",
    "strike": 14.5,
    "underlying": "CCL",
  },
  Object {
    "action": "Open",
    "amount": 6,
    "date": 2022-07-11T14:45:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.06,
    "quantity": 1,
    "side": "Sell",
    "strike": 3.5,
    "underlying": "CCL",
  },
  Object {
    "action": "Open",
    "amount": 10,
    "date": 2022-07-11T14:29:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.1,
    "quantity": 1,
    "side": "Sell",
    "strike": 47,
    "underlying": "INTC",
  },
  Object {
    "action": "Open",
    "amount": 20,
    "date": 2022-07-11T14:29:00.000Z,
    "expiration": "08/26/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.2,
    "quantity": 1,
    "side": "Sell",
    "strike": 28,
    "underlying": "INTC",
  },
  Object {
    "action": "Open",
    "amount": 26,
    "date": 2022-07-11T14:22:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.26,
    "quantity": 1,
    "side": "Sell",
    "strike": 30,
    "underlying": "ASAN",
  },
  Object {
    "action": "Open",
    "amount": 21,
    "date": 2022-07-08T18:51:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.21,
    "quantity": 1,
    "side": "Sell",
    "strike": 10,
    "underlying": "AAL",
  },
  Object {
    "action": "Open",
    "amount": 22,
    "date": 2022-07-08T18:51:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.22,
    "quantity": 1,
    "side": "Sell",
    "strike": 18,
    "underlying": "AAL",
  },
  Object {
    "action": "Open",
    "amount": 13,
    "date": 2022-07-08T18:43:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.13,
    "quantity": 1,
    "side": "Sell",
    "strike": 18,
    "underlying": "T",
  },
  Object {
    "action": "Open",
    "amount": 3,
    "date": 2022-07-08T18:43:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.03,
    "quantity": 1,
    "side": "Sell",
    "strike": 24,
    "underlying": "T",
  },
  Object {
    "action": "Open",
    "amount": 37,
    "date": 2022-07-08T15:41:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Put",
    "price": 0.37,
    "quantity": 1,
    "side": "Sell",
    "strike": 37.5,
    "underlying": "CSCO",
  },
  Object {
    "action": "Open",
    "amount": 15,
    "date": 2022-07-08T15:41:00.000Z,
    "expiration": "08/19/2022",
    "fees": 1.142,
    "optionType": "Call",
    "price": 0.15,
    "quantity": 1,
    "side": "Sell",
    "strike": 50,
    "underlying": "CSCO",
  },
]
`;
